### PR TITLE
[60] contributors can assign audiences

### PR DESCRIPTION
**Pull request by jazzsequence**
Previously, if logged in as a contributor, you could add a personalization block to a post, but you could not assign audiences to it. This was because the access that was being requested from the API for the audience post type was `edit` and contributors only have `view` permissions, therefore the API request returned a 403 forbidden error.

this PR updates a couple things that deal both with the functional changes as well as cosmetic limitations of the contributor user role:

* the `getPosts`, `getPost` and `getCurrentPost` functions now take an object of API query arguments so we can pass in `status` and `context` (and any other API query arguments we may need later) more easily when we're dealing with contributors. 
* we check for errors in `sortPosts` which may bubble up if posts were not able to be retrieved from the API, resulting in console errors
* the Audience list now checks whether the current user has the ability to create audiences, and passes that into the `getPosts` wrapper
* the up and down arrows in the Audience list are disabled if a user is not able to edit audiences
* the status toggle is removed if a user is not able to edit audiences

Screen Shot 2021-01-27 at 1 25 56 PM

fixes humanmade/experiments#60

### CODE REVIEW

**Comment by jazzsequence** [comment]
@roborourke let me know if you have any ideas off the top of your head about the bug above. I suspect the permissions may need to be adjusted for contributors for the `/analytics/v1/xbs` endpoint, but I haven't dug into it yet.

File: inc/audiences/namespace.php [hunk 1]
```diff
@@ -96,6 +96,7 @@ function register_post_type() {
 function maybe_grant_caps( $allcaps ) {
 	$cap_map = [
 		'edit_audiences' => 'edit_pages',
+		'read_audiences' => 'read',
```
Comment (jazzsequence) [inline inc/audiences/namespace.php:99]: I'm not sure if this is necessary or if it's implicit. However, as the post type is not public, perhaps it actually is necessary.

File: src/audiences/ui/data/reducer.js [hunk 2]
```diff
@@ -8,6 +8,9 @@ import { unionBy } from 'lodash';
  * @returns {number} Sort result value.
  */
 const sortPosts = ( a, b ) => {
+	if ( a.error || b.error ) {
```
Comment (jazzsequence) [inline src/audiences/ui/data/reducer.js:11]: This is possibly an unrelated fix, however it could be related to the bug that is happening now.

File: src/audiences/ui/list.js [hunk 3]
```diff
@@ -103,13 +103,19 @@ class List extends Component {
 	 */
 	onSearch = event => {
 		const value = event.target.value;
+		const canCreate = wp.data.select( 'core' ).canUser( 'create', 'audiences' );
```
Comment (jazzsequence) [inline src/audiences/ui/list.js:106]: I wasn't sure if there was a better/easier way to get `wp.data.select` but just using `select` here threw an error as it was not defined.

Comment (roborourke) [reply src/audiences/ui/list.js:106]: Use `this.props.canCreate()` here

File: src/audiences/ui/data/reducer.js (hunk 2 above)
Comment (roborourke) [reply src/audiences/ui/data/reducer.js:11]: We probably shouldn’t be adding posts that are actually errors to the list in the first place. I’d filter those out in the `getPosts()` function and maybe store or set an error in the redux store instead, then we could decide what to do with it / how to display it later

File: inc/audiences/namespace.php (hunk 1 above)
Comment (roborourke) [reply inc/audiences/namespace.php:99]: Best way to find out is to try removing it and see if everything still works

File: src/audiences/ui/data/reducer.js (hunk 2 above)
Comment (jazzsequence) [reply src/audiences/ui/data/reducer.js:11]: This isn't even in relation to posts. This appears to be any link in the element. It's not specific enough and it's erroring before the list is even displayed, hence at least filtering out the errored elements.

Case in point: I believe this causes an error in the "Select Audiences" link and the variant link in the block itself.

Comment (roborourke) [reply src/audiences/ui/data/reducer.js:11]: This sort posts function is just used when adding posts or updating posts in the redux store, it shouldn’t have any relation to the lower level components - I’m saying that errors should not be added to the list of posts in the first place. You could check the response in the getPosts function or in the reducer itself to make sure of that

Comment (jazzsequence) [reply src/audiences/ui/data/reducer.js:11]: ¯\_(ツ)_/¯ I will check it out but it seems to be affecting more than just the post list.

Comment (roborourke) [reply src/audiences/ui/data/reducer.js:11]: The posts are read from the list in the store so if we have an error stored instead of a post it would appear that way. I think in this instance it can be addressed at a lower level if that makes sense.

File: src/audiences/ui/list.js [hunk 4]
```diff
@@ -103,13 +103,19 @@ class List extends Component {
 	 */
 	onSearch = event => {
 		const value = event.target.value;
+		const canCreate = this.props.canCreate();
```
Comment (roborourke) [inline src/audiences/ui/list.js:106]: Not sure how this bit was working but `this.props.canCreate` is the value we passed through in the `applyWithSelect` higher order component. Here it's being called like a function when the value will be a boolean.

I'm suggesting removing this here and just checking `this.props.canCreate` below

```suggestion
```

File: src/audiences/ui/list.js [hunk 5]
```diff
@@ -103,13 +103,19 @@ class List extends Component {
...
 			search: value,
 		} );
 		// Query posts by the search term if we don't have an existing request.
 		if ( ! this.props.loading ) {
-			this.props.onGetPosts( this.state.page, value );
+			this.props.onGetPosts( {
+				page: this.state.page,
+				search: value,
+				context: canCreate ? 'edit' : 'view',
+				status: canCreate ? 'publish,draft' : 'publish',
```
Comment (roborourke) [inline src/audiences/ui/list.js:117]:
```suggestion
				context: this.props.canCreate ? 'edit' : 'view',
				status: this.props.canCreate ? 'publish,draft' : 'publish',
```

File: src/audiences/ui/list.js [hunk 6]
```diff
@@ -169,7 +175,10 @@ class List extends Component {
 	 */
 	onNextPage = () => {
 		const { page, search } = this.state;
-		this.props.onGetPosts( page + 1, search );
+		this.props.onGetPosts( {
+			page: page + 1,
+			search: search,
```
Comment (roborourke) [inline src/audiences/ui/list.js:180]: Presumably we need the same context and status checks parameters here
```suggestion
			search: search,
			context: this.props.canCreate ? 'edit' : 'view',
			status: this.props.canCreate ? 'publish,draft' : 'publish',
```

File: src/audiences/ui/list.js [hunk 7]
```diff
@@ -316,11 +325,13 @@ const applyWithSelect = withSelect( select => {
...
-	const posts = getPosts();
+	const canCreate = select( 'core' ).canUser( 'create', 'audiences' );
 	const loading = getIsLoading();
 	const pagination = getPagination();
-
-	const canCreate = select( 'core' ).canUser( 'create', 'audiences' );
+	const posts = canCreate ? getPosts() : getPosts( {
+		context: 'view',
+		status: 'publish',
+	} );
```
Comment (roborourke) [inline src/audiences/ui/list.js:334]: Ternary conditions tend to work best for shorter values, sometimes they have poor readability. It's a judgment call ultimately but for this logic I would probably write it like this in keeping with how you've done it elsewhere in this file.

```suggestion
	const queryArgs = {
		context: canCreate ? 'edit' : 'view',
		status: canCreate ? 'publish,draft' : 'publish',
	};
	const posts = getPosts( queryArgs );
```

What you could do is write a small helper function to get these params like:

```js
const getQueryArgs = ( queryArgs = {}, shouldOverride = false, overrides = {} ) => {
  return Object.assign(
    {
      context: 'view',
      status: 'publish',
    },
    queryArgs,
    shouldOverride ? overrides : {}
  );
}
```

Then use it like

```js
const posts = getPosts( getQueryArgs( {}, canCreate, {
  context: 'edit',
  status: 'publish,draft'
} ) );
```

Something like that anyway. It's not a blocker for me as the use case is limited to just this file so don't feel like you have to do that.

File: src/audiences/ui/select.js [hunk 8]
```diff
@@ -163,8 +163,12 @@ class Select extends Component {
 const applyWithSelect = withSelect( ( select, props ) => {
 	let audiencePost = null;
 
+	// Check if the current user can create audiences. If so, they should have create permissions sent to getPost.
+	const canCreate = select( 'core' ).canUser( 'create', 'audiences' ),
+		queryArgs = canCreate ? { context: 'edit' } : {};
```
Comment (roborourke) [inline src/audiences/ui/select.js:168]: I think the compiler might take care of optimisations like this now but it's not necessary to assign multiple values from a single `const` declaration. It causes problems with alignment anyway if spaces aren't used. It's usually work ok with `let` or `var` if your tab width was 4 spaces but here we're not gaining much.

```suggestion
	const canCreate = select( 'core' ).canUser( 'create', 'audiences' );
	const queryArgs = canCreate ? { context: 'edit' } : {};
```

**Review by roborourke — CHANGES_REQUESTED** [review]
Nice one, this is basically there. There are few small code style tweaks and 2 bugs where `canCreate` was being called like a function and the status & context args were missing from the `onNextPage` query args.

Good work pushing through those walls 👍

File: src/audiences/ui/components/list-row.js [hunk 9]
```diff
@@ -140,11 +140,17 @@ const ListRow = props => {
...
+				{ canEdit && (
+					<StatusToggle
+						status={ post.status }
+						onChange={ onStatusChange }
+					/>
+				) }
+				{ ! canEdit && post.status === 'publish' && (
+					<Fragment>
+						<span className="active">{ __( 'Active', 'altis-analytics' ) }</span>
+					</Fragment>
```
[the remark this answers is not in the document]
Comment (jazzsequence) [reply src/audiences/ui/components/list-row.js:152]: The opposite status isn't displayed for Contributors because they don't have access to draft posts they didn't create, and they don't have access to create audiences so I left it out.

File: src/audiences/ui/list.js (hunk 7 above)
Comment (jazzsequence) [reply src/audiences/ui/list.js:335]: I think the suggested change probably looks cleaner. I don't think we need a helper function at this point, but I've been wondering if something like it might be useful as I was going through these changes. I think it matters if this is something we come back to in the future and start using the query args more.

File: src/audiences/ui/select.js (hunk 8 above)
Comment (jazzsequence) [reply src/audiences/ui/select.js:168]: `¯\_(ツ)_/¯`

**Comment by jazzsequence** [comment]
This should be good for re-review

**Review by roborourke — APPROVED** [review]
If we’re happy it’s all still working for editors, authors and admin then let’s goooo 🚀

good work Chris
